gramps/gen/filters/optimizer.py

```diff
@@ -18,120 +19,86 @@
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 #
 
+from functools import reduce
 import logging
 
 LOG = logging.getLogger(".filter.optimizer")
 
 
 def intersection(sets):
-    if sets:
-        result = sets[0]
-        for s in sets[1:]:
-            result = result.intersection(s)
-        return result
-    else:
-        return set()
+    # sort the sets by length, shortest first.
+    # with interrsection, the shortest of the starting sets determines the
+    # maximum size of the result
+    sorted_sets = sorted(sets, key=lambda s: len(s))
+    return reduce(lambda x, y: x & y, sorted_sets[1:], sorted_sets[0])
```

Comment:
This is equally fast and perhaps a little easier to read:

```suggestion
    return set.intersection(*sorted_sets)
```

And then we also don't have to import/use functools.